Commiting Normalization Blog to main branch #3024 (#3231)
opensearch-project/project-website: _posts/2024-08-27-How-does-the-rank-normalization-work-in-hybrid-search.md

---
layout: post
title:  "An Overview of Rank Normalization in Hybrid Search"
authors:
 - ljeanniot

> date: 2024-08-30

Reviewer: Update to 2024-09-12

categories:
 - technical-post

> meta_keywords: semantic search engine, neural search engine, keyword and natural language search, hybrid search.

Reviewer: Please update the meta with the following: 
meta_keywords:  rank normalization, hybrid search, OpenSearch score normalization, search relevance, user experience

meta_description:  Learn about rank normalization in hybrid search and how to OpenSearch can help you standardize scores from diverse algorithms to deliver more accurate and fair search results.

Consider a hybrid search system that combines keyword-based search and vector-based search. Let's say we have the following raw scores from these sub-queries:

- **Keyword-Based Search Scores:** [2.0, 5.0, 3.0]
- **Vector-Based Search Scores:** [1.0, 4.0, 2.0]

**Application of Min-Max Normalization on These Scores:**

1. **Calculate the Minimum and Maximum Scores:**
   - Keyword-Based: Min = 2.0, Max = 5.0
   - Vector-Based: Min = 1.0, Max = 4.0

2. **Normalize the Scores:**
   - Keyword-Based: \([2.0-2.0]/[5.0-2.0], [5.0-2.0]/[5.0-2.0], [3.0-2.0]/[5.0-2.0] = [0.0, 1.0, 0.33]\)
   - Vector-Based: \([1.0-1.0]/[4.0-1.0], [4.0-1.0]/[4.0-1.0], [2.0-1.0]/[4.0-1.0] = [0.0, 1.0, 0.33]\)

**Visualization of the Scores Before and After Normalization:**

**Before Normalization:**
- Keyword-Based: [2.0, 5.0, 3.0]
- Vector-Based: [1.0, 4.0, 2.0]

**After Min-Max Normalization:**
- Keyword-Based: [0.001, 1.0, 0.33]
- Vector-Based: [0.001, 1.0, 0.33]

By applying Min-Max normalization, we ensure that the scores from different sub-queries are on a comparable scale, allowing for a fair combination and ranking of search results.

### 4. Comparing L2 and Min-Max Normalization

> **Advantages and Disadvantages:**

Reviewer 2: ```suggestion
```

**Benefits of Using L2 Normalization:**
1. **Smooth Adjustment:** L2 normalization provides a smooth adjustment of scores based on the overall distribution, ensuring that no single score disproportionately affects the result.
2. **Magnitude Preservation:** By taking into account the magnitude of all scores, L2 normalization effectively preserves the relative importance of each score in the context of the entire dataset.

> 3. **Handling Outliers:** L2 normalization can mitigate the effect of outliers, as extreme values have less impact on the final normalized scores compared to Min-Max normalization.

Reviewer 2: ```suggestion
- **Handling outliers:** L2 normalization can mitigate the effect of outliers because extreme values have less impact on the final normalized scores as compared to min-max normalization.
```

**Benefits of Using Min-Max Normalization:**
1. **Simple Interpretation:** Min-Max normalization scales scores to a fixed range (typically 0 to 1), making it easy to interpret and compare scores directly.
2. **Uniform Scale:** This method ensures that all scores fall within the same scale, which can be useful in applications where a consistent range is required.
3. **Extreme Value Emphasis:** Min-Max normalization highlights the importance of the minimum and maximum values, which can be beneficial in scenarios where the relative ranking between the lowest and highest scores is crucial.

> **Potential Drawbacks and When to Use Each Method:**

Reviewer 2: ```suggestion
**Potential drawbacks and when to use each method**
```

> 

Reviewer 2: Above heading: H3?

**L2 Normalization Drawbacks:**
1. **Complex Interpretation:** The resulting scores from L2 normalization can be less intuitive to interpret compared to Min-Max normalized scores.
2. **Computational Overhead:** Calculating the L2 norm requires additional computational effort, which might not be ideal for large datasets or real-time applications.

> **Min-Max Normalization Drawbacks:**

Reviewer 2: ```suggestion
**Min-max normalization drawbacks**
```

1. **Sensitive to Outliers:** Min-Max normalization can be significantly affected by outliers, as extreme values can disproportionately influence the normalized scores.
2. **Loss of Relative Magnitude:** This method does not preserve the relative magnitude of scores as effectively as L2 normalization, potentially leading to a loss of contextual information.

**When to Use Each Method:**

> - **L2 Normalization:** This method is preferred when the goal is to maintain the relative importance of each score within the context of the entire dataset and when the presence of outliers needs to be minimized.

Reviewer 2: ```suggestion
- **L2 normalization:** This method is preferred when the goal is to maintain the relative importance of each score within the context of the entire dataset and when the presence of outliers needs to be minimized.
```

- **Min-Max Normalization:** This method is ideal for applications that require a simple, fixed range for scores and where the extreme values need to be emphasized.

**Use Cases:**

> **Scenarios Where L2 Normalization is More Effective:**

Reviewer 2: ```suggestion
**Scenarios in which L2 normalization is more effective**
```

> 1. **Document Ranking in Search Engines:** When ranking documents based on multiple relevance factors, L2 normalization ensures that no single factor disproportionately influences the final ranking.

Reviewer 2: ```suggestion
- **Document ranking in search engines:** When ranking documents based on multiple relevance factors, L2 normalization ensures that no single factor disproportionately influences the final ranking.
```

> 2. **Recommender Systems:** In scenarios where recommendations are based on multiple user preferences, L2 normalization helps balance the different preferences without allowing any single preference to dominate.

Reviewer 2: ```suggestion
- **Recommender systems:** In scenarios where recommendations are based on multiple user preferences, L2 normalization helps balance the different preferences without allowing any single preference to dominate.
```

3. **Machine Learning Feature Scaling:** When preparing data for machine learning models, L2 normalization can be used to scale features while preserving their relative importance.

> **Scenarios Where Min-Max Normalization is Preferred:**

Reviewer 2: ```suggestion
**Scenarios in which min-max normalization is preferred**
```

> 1. **Image Processing:** In image processing tasks, Min-Max normalization is commonly used to scale pixel values to a standard range (e.g., 0 to 1) for consistency and ease of interpretation.

Reviewer 2: ```suggestion
- **Image processing:** In image processing tasks, min-max normalization is commonly used to scale pixel values to a standard range (for example, 0 to 1) for consistency and ease of interpretation.
```

> 2. **Financial Data Analysis:** When analyzing financial data, such as stock prices or returns, Min-Max normalization can be used to compare different datasets on a common scale.

Reviewer 2: ```suggestion
- **Financial data analysis:** When analyzing financial data, such as stock prices or returns, min-max normalization can be used to compare different datasets on a common scale.
```

3. **User Ratings:** In systems where user ratings are combined from different sources, Min-Max normalization ensures that all ratings are scaled to a consistent range, making it easier to compare and aggregate them.

### 5. Implementation in OpenSearch

**Overview of How These Normalization Techniques are Implemented in OpenSearch:**

OpenSearch implements normalization techniques via search pipelines, processing search results to standardize scores from different sub-queries. This ensures accurate aggregation and ranking of results, supporting various normalization methods like L2 and Min-Max.

> **Configuration and Usage in an OpenSearch Environment:**

Reviewer 2: ```suggestion
**Configuration and usage in an OpenSearch environment**
```

> 1. **Define the Search Pipeline:**

Reviewer 2: ```suggestion
1. **Define the search pipeline:** Create a search pipeline with a normalization processor and specify the normalization technique:
```

   - Create a search pipeline with a normalization processor and specify the normalization technique.

```json
PUT /_search/pipeline/nlp-search-pipeline
{
  "description": "Post processor for hybrid search",
  "phase_results_processors": [
    {
      "normalization-processor": {
        "normalization": {
          "technique": "min_max"
        },
        "combination": {
          "technique": "arithmetic_mean",
          "parameters": {
            "weights": [
              0.3,
              0.7
            ]
          }
        }
      }
    }
  ]
}
```

> 2. **Utilize the Search Pipeline:**

Reviewer 2: ```suggestion
2. **Use the search pipeline:** Use the search pipeline in search requests to apply normalization and combination techniques to search results:
```

>    - Use the search pipeline in search requests to apply normalization and combination techniques to search results.

Reviewer 2: ```suggestion
```

```json
GET /my-nlp-index/_search?search_pipeline=nlp-search-pipeline
{
  "_source": {
    "exclude": [
      "passage_embedding"
    ]
  },
  "query": {
    "hybrid": {
      "queries": [
        {
          "match": {
            "text": {
              "query": "horse"
            }
          }
        },
        {
          "neural": {
            "passage_embedding": {
              "query_text": "wild west",
              "model_id": "aVeif4oB5Vm0Tdw8zYO2",
              "k": 5
            }
          }
        }
      ]
    }
  }
}
```

> **Practical Tips for Developers to Optimize Search Results Using Normalization Techniques:**

Reviewer 2: ```suggestion
**Practical tips for developers to optimize search results using normalization techniques**
```

> 1. **Choose the Right Technique:**

Reviewer 2: ```suggestion
1. **Choose the right technique:** Assess your data and query nature to select the most suitable normalization technique, such as L2 for varying magnitudes or min-max for fixed-range scaling.
```

   - Assess your data and query nature to select the most suitable normalization technique, such as L2 for varying magnitudes and Min-Max for fixed range scaling.

2. **Test and Tune:**

>    - Continuously test normalized search results to ensure they meet relevance and ranking criteria. Adjust the pipeline configuration as necessary.

Reviewer 2: ```suggestion
```

> 3. **Monitor Performance:**

Reviewer 2: ```suggestion
3. **Monitor performance:** Track the performance impact of normalization, especially in high-traffic environments. Optimize configurations to balance accuracy and performance.
```

>    - Track the performance impact of normalization, especially in high-traffic environments. Optimize configurations to balance accuracy and performance.

Reviewer 2: ```suggestion
```

4. **Handle Edge Cases:**
   - Prepare for edge cases like identical scores by ensuring normalization logic handles these scenarios without skewing results. Some possible edge cases include:

>      - **Identical Scores:** Implement logic to handle cases where all scores are identical to avoid division by zero or uniform scores.

Reviewer 2: ```suggestion
- **Identical scores:** Implement logic to handle cases in which all scores are identical to avoid division by zero or uniform scores.
```

>      - **Zero Scores:** Introduce a small epsilon value to avoid issues with zero scores in calculations.

Reviewer 2: ```suggestion
- **Zero scores:** Introduce a small epsilon value to avoid issues with zero scores in calculations.
```

>      - **Sparse Data:** Address imbalanced data distributions to prevent skewed normalization results.

Reviewer 2: ```suggestion
- **Sparse data:** Address imbalanced data distributions to prevent skewed normalization results.
```

>      - **Outliers:** Apply techniques to minimize the impact of extreme values on the normalized scores.

Reviewer 2: ```suggestion
- **Outliers:** Apply techniques to minimize the impact of extreme values on the normalized scores.
```

>      - **Empty Sub-Queries:** Ensure empty sub-queries are handled gracefully without affecting the overall normalization process.

Reviewer 2: ```suggestion
- **Empty sub-queries:** Ensure empty sub-queries are handled gracefully without affecting the overall normalization process.
```

> 5. **Leverage Documentation:**

Reviewer 2: ```suggestion
5. **Leverage documentation:** Leverage OpenSearch documentation and community resources for additional guidance on implementing and optimizing normalization techniques.
```

>    - Utilize OpenSearch documentation and community resources for additional guidance on implementing and optimizing normalization techniques.

Reviewer 2: ```suggestion
```

By following these steps, developers can effectively leverage normalization techniques in OpenSearch to deliver accurate and fair search results, enhancing the overall search experience for users. For more detailed information, refer to the [OpenSearch documentation on normalization processors](https://opensearch.org/docs/latest/search-plugins/search-pipelines/normalization-processor/).

### Conclusion

Rank normalization is crucial in hybrid search systems, ensuring scores from different algorithms are comparable for accurate and fair search results. By using techniques like L2 and Min-Max normalization, OpenSearch provides tools to standardize scores effectively. Experimenting with these methods helps identify the best fit for specific search needs, optimizing relevance and user satisfaction.

We encourage you to explore both normalization techniques and share your experiences. Your feedback and discussions on rank normalization strategies are invaluable in refining and advancing search technologies.

> ### References and Further Reading

Reviewer 2: ```suggestion
### References and further reading
```

For more information on hybrid search, normalization processors, and implementation details, check out the following resources:

> 1. [Hybrid Search in OpenSearch](https://opensearch.org/docs/latest/search-plugins/hybrid-search/)

Reviewer 2: ```suggestion
1. [Hybrid search](https://opensearch.org/docs/latest/search-plugins/hybrid-search/)
```

> 2. [Normalization Processor in OpenSearch](https://opensearch.org/docs/latest/search-plugins/search-pipelines/normalization-processor/)

Reviewer 2: ```suggestion
2. [Normalization processor](https://opensearch.org/docs/latest/search-plugins/search-pipelines/normalization-processor/)
```
